homeassistant/components/zwave_js/update.py

```diff
@@ -0,0 +1,198 @@
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_firmware_update_entity",
+            async_add_firmware_update_entity,
+        )
+    )
+
+
+def get_latest_firmware(
+    firmwares: list[FirmwareUpdateInfo],
+) -> FirmwareUpdateInfo:
+    """Get the latest firmware from the available updates."""
+    latest_firmware = FirmwareUpdateInfo("0.0.0", "", [])
+    for firmware in firmwares:
+        if AwesomeVersion(firmware.version) > AwesomeVersion(latest_firmware.version):
+            latest_firmware = firmware
+    return latest_firmware
+
+
+class ZWaveNodeFirmwareUpdate(UpdateEntity):
+    """Representation of a firmware update entity."""
+
+    _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = UpdateDeviceClass.FIRMWARE
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL
+        | UpdateEntityFeature.RELEASE_NOTES
+        | UpdateEntityFeature.SPECIFIC_VERSION
+    )
+    _attr_has_entity_name = True
+
+    def __init__(self, driver: Driver, node: ZwaveNode) -> None:
+        """Initialize a Z-Wave device firmware update entity."""
+        self.driver = driver
+        self.node = node
+        self.available_firmware_updates: list[FirmwareUpdateInfo] = []
+        self._latest_version_firmware: FirmwareUpdateInfo | None = None
+
+        # Entity class attributes
+        self._attr_name = "Firmware"
+        self._base_unique_id = get_valueless_base_unique_id(driver, node)
+        self._attr_unique_id = f"{self._base_unique_id}.firmware_update"
+        # device may not be precreated in main handler yet
+        self._attr_device_info = DeviceInfo(
+            identifiers={get_device_id(driver, node)},
+            sw_version=node.firmware_version,
+            name=node.name or node.device_config.description or f"Node {node.node_id}",
+            model=node.device_config.label,
+            manufacturer=node.device_config.manufacturer,
+            suggested_area=node.location if node.location else None,
+        )
+
+        self._attr_installed_version = self._attr_latest_version = node.firmware_version
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """Return entity specific state attributes."""
+
+        return {
+            ATTR_AVAILABLE_FIRMWARE_UPDATES: [
+                {
+                    "version": firmware.version,
+                    "release_notes": firmware.changelog,
+                }
+                for firmware in sorted(
+                    self.available_firmware_updates,
+                    key=lambda f: AwesomeVersion(f.version),
+                )
+            ]
+        }
+
+    async def async_update(self) -> None:
+        """Update the entity."""
+        self.available_firmware_updates = (
+            await self.driver.controller.async_get_available_firmware_updates(
```

Comment:
Can we check for sleeping device first? Otherwise it will be pretty common that this fails.

```
2022-08-29 14:45:08.052 ERROR (MainThread) [homeassistant.components.update] zwave_js: Error on device update!
Traceback (most recent call last):
  File "/home/martin/dev/home-assistant/core/homeassistant/helpers/entity_platform.py", line 477, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/home/martin/dev/home-assistant/core/homeassistant/helpers/entity.py", line 703, in async_device_update
    await task
  File "/home/martin/dev/home-assistant/core/homeassistant/components/zwave_js/update.py", line 123, in async_update
    await self.driver.controller.async_get_available_firmware_updates(
  File "/home/martin/.virtualenvs/home-assistant/lib/python3.9/site-packages/zwave_js_server/model/controller/__init__.py", line 719, in async_get_available_firmware_updates
    data = await self.client.async_send_command(
  File "/home/martin/.virtualenvs/home-assistant/lib/python3.9/site-packages/zwave_js_server/client.py", line 103, in async_send_command
    return await future
zwave_js_server.exceptions.FailedZWaveCommand: Z-Wave error 260: Cannot check for firmware updates for node 10: The node did not wake up within 1 minute! (ZW0260)
```

Reply:
To limit the parallel updates, you can use a shared [semaphore](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore).

```python
async with self.driver.update_semaphore:
    self.available_firmware_updates = await ...
```

Reply:
btw. I think that it's kinda wild that this command queries the actual device for it's current firmware? I would expect that to remain cached and maybe queried every time we first see it (on startup, whatever). This should only query the FW server

Reply:
But this is not doing an update right? This is only fetching the latest info from the update server

Reply:
correct